### PR TITLE
[rocprofiler-compute] Add method-aware --pc-sampling-interval default

### PR DESCRIPTION
**Pull request by vedithal-amd**
## Motivation

`--pc-sampling-interval` previously defaulted to `1048576` for both
sampling methods. That value suits `stochastic` (cycles) but is far too
large for `host_trap` (microseconds), making the default profiling run
ineffective. This makes the default method-aware and validates
user-supplied values.

Stacked on #6890; the base of this PR is that branch, not
`rocprofiler-compute-develop`. Retarget once #6890 merges.

## Technical Details

- `--pc-sampling-interval` defaults to `None` in argparse and resolves in
  `sanitize()` to a method-appropriate value: 512 microseconds for
  `host_trap`, 1048576 cycles for `stochastic`, so every downstream
  consumer reads the concrete value.
- Bugfix: stochastic intervals are validated to be a power of 2 and at
  least 65536, and host_trap intervals to be positive; invalid values now
  raise a clear error instead of being passed through silently.
- `type=int` on the argument rejects non-integer input at parse time.

## JIRA ID

AIPROFCOMP-605

## Test Plan

- `pytest tests/test_profiler_base.py`
- Smoke: `rocprof-compute profile --experimental --pc-sampling
  --pc-sampling-method host_trap ... -- app` records a 512 us interval;
  `--pc-sampling-method stochastic --pc-sampling-interval 12345` is
  rejected.

## Test Result

- [x] `ruff check` / `ruff format --check` clean
- [x] `pytest tests/test_profiler_base.py`: 37 passed
- [ ] GPU-dependent PC sampling profile tests left to CI

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-compute/src/rocprof_compute_base.py [hunk 1]
```diff
@@ -629,6 +632,40 @@ def _validate_profile_mode_arguments(self) -> None:
...
 
+        self._resolve_pc_sampling_interval()
+
+    def _resolve_pc_sampling_interval(self) -> None:
+        """Apply the method-aware default for --pc-sampling-interval and
+        validate a user-supplied stochastic interval.
+
+        stochastic intervals are in cycles and must be a power of two no
+        smaller than 65536; host_trap intervals are in microseconds and may
+        be any positive integer.
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/src/rocprof_compute_base.py]: Second stanza is redundant - please tighten the docstring

File: projects/rocprofiler-compute/src/rocprof_compute_base.py [hunk 2]
```diff
@@ -629,6 +632,40 @@ def _validate_profile_mode_arguments(self) -> None:
         if getattr(args, "bench_only", False) and getattr(args, "no_roof", False):
             console_error("--bench-only cannot be used with --no-roof.")
 
+        self._resolve_pc_sampling_interval()
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/src/rocprof_compute_base.py]: The docstring of this function means that this is not a appropriate parent - move it one level up the stakc with proifle guard like other cehcks

File: projects/rocprofiler-compute/src/rocprof_compute_base.py [hunk 3]
```diff
@@ -51,6 +51,9 @@
 from utils.utils_exceptions import WorkloadCommandError
 from utils.utils_profile import get_submodules
 
+PC_SAMPLING_INTERVAL_DEFAULTS = {"stochastic": 1048576, "host_trap": 512}
+PC_SAMPLING_STOCHASTIC_MIN_INTERVAL = 65536
+
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/src/rocprof_compute_base.py]: These should be function local vars

File: projects/rocprofiler-compute/tests/test_profiler_base.py [hunk 4]
```diff
@@ -297,6 +297,8 @@ def _make_rpc_args(
     filter_blocks=None,
     filter_metrics=None,
     pc_sampling=False,
+    pc_sampling_method="stochastic",
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profiler_base.py:309]: I know its a out of scope but i would rather solve it here - bvut can we move this function on the top and get rid of _make_sanitize_* in favor of this function and update its callers. Also, can we rename _make_rpc_with_args to just _make_rpc

File: projects/rocprofiler-compute/src/rocprof_compute_base.py (hunk 1 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py]: Tightened the docstring to a single sentence; dropped the per-method detail stanza.

File: projects/rocprofiler-compute/src/rocprof_compute_base.py (hunk 2 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py]: Moved the call into sanitize()'s profile-mode guard alongside the other checks; _validate_profile_mode_arguments no longer invokes it.

File: projects/rocprofiler-compute/src/rocprof_compute_base.py (hunk 3 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py]: Done, the defaults dict and the stochastic minimum are now locals inside the helper.

File: projects/rocprofiler-compute/tests/test_profiler_base.py (hunk 4 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/tests/test_profiler_base.py:309]: Done. _make_rpc_args is now the single builder at the top (superset namespace for both sanitize paths), _make_sanitize_args is gone and its callers updated, and _make_rpc_with_args is renamed _make_rpc.

**Comment by feizheng10** [comment]
Could you show some proof for " That value suits stochastic (cycles) but is far too large for host_trap (microseconds)" ?

**Comment by vedithal-amd** [comment]
@feizheng10 Two parts: a measurement, and the reasoning behind the new 512 us value.

**Measurement.** Profiling `tests/rocflop` on MI300X with `host_trap`, identical except for the interval:

| interval (us) | samples collected | samples/sec |
|---|---:|---:|
| 1048576 | 2,185 | 226 |
| 512 | 4,178,604 | 202,226 |

At 1048576 us the run is ~1900x sparser. In the analyze per-instruction view nearly every sampled instruction has a hit count of 1, so per-line attribution is statistical noise. (It does still collect some samples, not zero, because each host_trap poll drains the buffer rather than capturing exactly one.)

**Why 512.** The interval unit differs by method: cycles for stochastic, microseconds for host_trap. The old shared default of 1048576 is fine as cycles but means ~1.05 s between samples when read as microseconds, hence the ~2000x gap.

512 us is chosen to match the cadence the stochastic default already implies. 1048576 cycles at MI300X's 2.1 GHz shader clock is ~499 us, so 512 us (2^9) gives host_trap roughly the same ~0.5 ms sampling interval in the time domain. This also holds across the supported GPUs, since peak shader clock only spans ~1.7-2.4 GHz from MI250X to MI355X: 1048576 cycles is ~437-617 us across that range, and 512 us sits in the middle, so a single fixed default stays well-matched without needing to be per-arch.

File: projects/rocprofiler-compute/src/rocprof_compute_base.py [hunk 5]
```diff
@@ -629,6 +630,36 @@ def _validate_profile_mode_arguments(self) -> None:
...
+        if not getattr(args, "pc_sampling", False):
+            return
+
+        interval_defaults = {"stochastic": 1048576, "host_trap": 512}
+        stochastic_min_interval = 65536
+
+        method = args.pc_sampling_method
+        if args.pc_sampling_interval is None:
+            args.pc_sampling_interval = interval_defaults[method]
+            return
```
[the remark this answers is not in the document]
Comment (feizheng10) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py:648]: If --pc-sampling-method is not restricted in argparse, _resolve_pc_sampling_interval() does interval_defaults[method] and can KeyError on a typo. Prefer choices= on the flag or an explicit unknown-method error before the dict lookup.

**Review by feizheng10 — APPROVED** [review]
Maybe add tests for host_trap with 0 / negative interval.
Just some minor comments. Apporved.

File: projects/rocprofiler-compute/docs/how-to/pc_sampling.rst [hunk 6]
```diff
@@ -28,13 +28,13 @@ Profiling options
 For using profiling options for PC sampling the configuration needed are:
 
 * ``--pc-sampling-method``: Should be either ``stochastic`` or ``host_trap``, (DEFAULT: stochastic)
-* ``--pc-sampling-interval``: For stochastic sampling, the interval is in cycles. The finest granularity is 1 cycle. For ``host_trap`` sampling, the interval is in microsecond (DEFAULT: 1048576). The interval should be the power of 2. You are recommended try starting from 1048576, and lowering until reaching 65536.
+* ``--pc-sampling-interval``: For ``stochastic`` sampling, the interval is in cycles; it must be a power of 2 and at least 65536 (DEFAULT: 1048576). For ``host_trap`` sampling, the interval is in microseconds and may be any positive integer (DEFAULT: 512). When omitted, the method-appropriate default is used.
```
Comment (xuchen-amd) [inline projects/rocprofiler-compute/docs/how-to/pc_sampling.rst]:
```
``stochastic``... it must be a power of 2 and at least 65536 (DEFAULT: 1048576)
```

should we add an explanation in the docs to why this is needed?

File: projects/rocprofiler-compute/docs/how-to/pc_sampling.rst [hunk 7]
```diff
@@ -28,13 +28,13 @@ Profiling options
...
 
 **Sample command:**
 
 .. code-block:: shell
 
-   $ rocprof-compute profile -n pc_test --no-roof --experimental --pc-sampling --pc-sampling-method stochastic --pc-sampling-interval 1048576 -VVV -- target_app
+   $ rocprof-compute profile -n pc_test --experimental --pc-sampling --no-roof --pc-sampling-method stochastic -VVV -- target_app
```
Comment (xuchen-amd) [inline projects/rocprofiler-compute/docs/how-to/pc_sampling.rst]:
```suggestion
   $ rocprof-compute profile -n pc_test --no-roof --experimental --pc-sampling --pc-sampling-method stochastic -VVV -- target_app
```

Not sure if you need to update the commit history, I believe this was fixed in part 1.

File: projects/rocprofiler-compute/tests/test_pc_sampling.py
Comment (xuchen-amd) [inline projects/rocprofiler-compute/tests/test_pc_sampling.py]: should this be updated/removed?

Comment (xuchen-amd) [inline projects/rocprofiler-compute/tests/test_pc_sampling.py]: should this be updated/removed?

Comment (xuchen-amd) [inline projects/rocprofiler-compute/tests/test_pc_sampling.py]: should this be updated/removed?

File: projects/rocprofiler-compute/src/rocprof_compute_base.py [hunk 8]
```diff
@@ -629,6 +630,36 @@ def _validate_profile_mode_arguments(self) -> None:
...
 
+    def _resolve_pc_sampling_interval(self) -> None:
+        """Apply the method-aware default for --pc-sampling-interval and
+        validate a user-supplied value."""
+        args = self.__args
+        if not getattr(args, "pc_sampling", False):
+            return
+
+        interval_defaults = {"stochastic": 1048576, "host_trap": 512}
+        stochastic_min_interval = 65536
```
Comment (xuchen-amd) [inline projects/rocprofiler-compute/src/rocprof_compute_base.py]: could we do something like the below (example only) for clarity?
```suggestion
        stochastic_default_interval_in_cycles = 1048576
        stochastic_min_interval_in_cycles = 65536
        host_trap_default_interval_in_ms = 512
```

File: projects/rocprofiler-compute/src/rocprof_compute_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (xuchen-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py:648]: seconding this comment and prefer fix to be in arg parser.

**Review by xuchen-amd — CHANGES_REQUESTED** [review]
Left some comments and questions, thanks!

File: projects/rocprofiler-compute/src/rocprof_compute_base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py:648]: Fixed in arg parser: `--pc-sampling-method` now has `choices=["stochastic", "host_trap"]`, so an unknown method is rejected at parse time before `_resolve_pc_sampling_interval()` runs. This closes both the `KeyError` on the defaults lookup and the silent host_trap fallthrough.

[the remark this answers is not in the document]
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py:648]: Fixed in arg parser: `--pc-sampling-method` now has `choices=["stochastic", "host_trap"]`, so an unknown method is rejected at parse time before `_resolve_pc_sampling_interval()` runs. This closes both the `KeyError` on the defaults lookup and the silent host_trap fallthrough.

[the remark this answers is not in the document]
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py:648]: Fixed in arg parser: `--pc-sampling-method` now has `choices=["stochastic", "host_trap"]`, so an unknown method is rejected at parse time before `_resolve_pc_sampling_interval()` runs. This closes both the `KeyError` on the defaults lookup and the silent host_trap fallthrough.

File: projects/rocprofiler-compute/tests/test_profiler_base.py [hunk 9]
```diff
@@ -390,10 +386,41 @@ def _make_rpc_with_args(args: argparse.Namespace) -> RocProfCompute:
...
+@pytest.mark.parametrize(
+    "method, interval, expect_error, expected_interval",
+    [
+        pytest.param("host_trap", None, False, 512, id="host_trap_unset_default"),
+        pytest.param("stochastic", None, False, 1048576, id="stochastic_unset_default"),
+        pytest.param("stochastic", 65536, False, 65536, id="stochastic_min_accepted"),
+        pytest.param("stochastic", 12345, True, None, id="stochastic_not_pow2"),
+        pytest.param("stochastic", 32768, True, None, id="stochastic_below_min"),
+        pytest.param("host_trap", 100, False, 100, id="host_trap_positive_accepted"),
+    ],
```
[the remark this answers is not in the document]
Comment (vedithal-amd) [reply projects/rocprofiler-compute/tests/test_profiler_base.py:558]: Added `host_trap_zero_rejected` (interval 0) and `host_trap_negative_rejected` (interval -1) cases to `test_sanitize_pc_sampling_interval`, both asserting `SystemExit`, so the `interval <= 0` path is now exercised.

File: projects/rocprofiler-compute/tests/test_pc_sampling.py
Comment (vedithal-amd) [reply projects/rocprofiler-compute/tests/test_pc_sampling.py]: Removed. These host_trap tests now rely on the method-aware default (512 us) like the other PC sampling tests, so the explicit `--pc-sampling-interval 256` is gone.

Comment (vedithal-amd) [reply projects/rocprofiler-compute/tests/test_pc_sampling.py]: Removed. These host_trap tests now rely on the method-aware default (512 us) like the other PC sampling tests, so the explicit `--pc-sampling-interval 256` is gone.

Comment (vedithal-amd) [reply projects/rocprofiler-compute/tests/test_pc_sampling.py]: Removed. These host_trap tests now rely on the method-aware default (512 us) like the other PC sampling tests, so the explicit `--pc-sampling-interval 256` is gone.

File: projects/rocprofiler-compute/docs/how-to/pc_sampling.rst (hunk 6 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/docs/how-to/pc_sampling.rst]: Added a one-line note that the power-of-2 / 65536 stochastic constraints are hardware limits reported by the driver.

File: projects/rocprofiler-compute/src/rocprof_compute_base.py (hunk 8 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/src/rocprof_compute_base.py]: Done, with unit-suffixed names: `stochastic_default_interval_in_cycles`, `stochastic_min_interval_in_cycles`, and `host_trap_default_interval_in_microseconds` (host_trap is microseconds, not ms).

File: projects/rocprofiler-compute/docs/how-to/pc_sampling.rst (hunk 7 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/docs/how-to/pc_sampling.rst]: Done, reordered the sample command flags to match: `-n pc_test --no-roof --experimental --pc-sampling --pc-sampling-method stochastic -VVV -- target_app`.

**Review by xuchen-amd — APPROVED** [review]
Thanks for addressing my comments!
